libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h

```diff
@@ -0,0 +1,172 @@
+    {
+#  ifndef _LIBCUDACXX_NO_EXCEPTIONS
+      throw ::cuda::cuda_error{cudaErrorInvalidValue, "Invalid alignment passed to cuda_memory_resource::allocate."};
+#  else
+      _LIBCUDACXX_UNREACHABLE();
+#  endif
+    }
+
+    void* __ptr{nullptr};
+    const ::cudaError_t __status = ::cudaMalloc(&__ptr, __bytes);
+    switch (__status)
+    {
+      case ::cudaSuccess:
+        return __ptr;
+      default:
+        ::cudaGetLastError(); // Clear CUDA error state
+#  ifndef _LIBCUDACXX_NO_EXCEPTIONS
+        throw ::cuda::cuda_error{__status, "Failed to allocate memory with cudaMalloc."};
+#  else
+        _LIBCUDACXX_UNREACHABLE();
+#  endif
+    }
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p __ptr.
+   * @param __ptr Pointer to be deallocated. Must have been allocated through a call to `allocate`
+   * @param __bytes The number of bytes that was passed to the `allocate` call that returned \p __ptr.
+   * @param __alignment The alignment that was passed to the `allocate` call that returned \p __ptr.
+   */
+  void deallocate(void* __ptr, const size_t __bytes, const size_t __alignment = __default_cuda_malloc_alignment) const
+  {
+    // We need to ensure that the provided alignment matches the minimal provided alignment
+    _LIBCUDACXX_ASSERT(
+      __default_cuda_malloc_alignment <= __alignment && (__alignment % __default_cuda_malloc_alignment == 0),
+      "Invalid alignment passed to cuda_memory_resource::deallocate.");
+    const ::cudaError_t __status = ::cudaFree(__ptr);
+    (void) __status;
+    _LIBCUDACXX_ASSERT(__status == cudaSuccess, "cuda_memory_resource::deallocate failed");
+  }
+
+  /**
+   * @brief Equality comparison with another cuda_memory_resource
+   * @return true
+   */
+  _LIBCUDACXX_NODISCARD_ATTRIBUTE constexpr bool operator==(cuda_memory_resource const&) const noexcept
+  {
+    return true;
+  }
+#  if _CCCL_STD_VER <= 2017
+  /**
+   * @brief Inequality comparison with another cuda_memory_resource
+   * @return false
+   */
+  _LIBCUDACXX_NODISCARD_ATTRIBUTE constexpr bool operator!=(cuda_memory_resource const&) const noexcept
+  {
+    return false;
+  }
+#  endif // _CCCL_STD_VER <= 2017
+
+  /**
+   * @brief Equality comparison between a cuda_memory_resource and another resource
+   *
+   * We want to ensure that we can compare any cuda_memory_resource too a resource_ref that holds a cuda_memory_resource
+   * Therefore, we convert both to a resource_ref<> which is the lowest common denominator. The resource_ref<>
+   * internally checks whether both resources are of the same kind and if so returns the comparison of the two.
+   * Otherwise, the comparison will always return false, which is what we want for any other memory resource.
+   *
+   * @param __lhs The cuda_memory_resource
+   * @param __rhs The resource to compare to
+   * @return Comparison of both resources converted to a resource_ref<>
```

Comment:
Suggestion: The semantics are pretty simple, can we just spell it out something like the following (or whatever the incantation would be)?

```suggestion
   * @return true if both arguments are convertible to resource_ref<>(cuda_memory_resource&), false otherwise
```

Reply:
I went with: Result of equality comparison of both resources converted to a resource_ref<>

Reply:
OK, but then the reader has to go dig up docs on the semantics of comparison of refs. Doesn't that just compare the referenced object types? Is there a section in the docs on equality comparison that you could link to?

std::pmr::memory_resource and RMM (current) memory resources have very different comparison semantics, so I think this should be clearly documented.